
Welcome to scPower, the statistical framework for design and power analysis of multi-sample single cell genomics experiments!</br>
The tool supports the user to set the experimental parameters of cell type specific inter-individual DE and eQTL analysis using single cell RNA-seq data. Experimental design suggestions are made in a way to optimize the power of the experiment.</br>
Thereby, the power is split into two subcategories:

<ul>
    <li>Power to detect DE/eQTL genes</li>
    <li>Power to detect cell types</li>
</ul>

For each power, an own section exists in the header of this website. A short explanation for each is given in the following sections. The website is only designed for 10X Genomics experiments. However, the model itself, defined in the R package scPower (see section "References"), is more flexible, suitable for many different single cell RNA-seq technologies, for example Drop-seq and Smart-seq.

<h4>Power to detect DE/eQTL genes</h4>
The "Power to detect DE/eQTL genes" evaluates the parameter combination which maximizes the detection power of DE / eQTL genes. The main plot on the right side shows a grid of parameter combinations. Always two parameter axis are defined by the user, the third parameter is defined uniquely given the other two parameters and the overall budget. The color visualizes the detection power of each combination. Depending on the overall budget, not all parameter combinations will be possible and some spots will stay white in the grid. A arrow called "selected study" points on the study with the highest detection power and the two plots below visualize the power curves for this study. The arrow can be set to any parameter combination by clicking on the main plot.</br>

The detection power is the product of the expression probability and the DE/eQTL power. The expression probability shows how likely it is that the DE/eQTL genes are expressed, while the DE/eQTL power shows how likely it is to detect the genes as significant, given that they are expressed. The two plots below the main plot show the influence of the parameters on each of the probabilities. The plot on the left depicts the influence of the parameter on the x axis, while the parameter on the y axis is kept constant (taking the value of the selected study). The plots on the right shows the same for the parameter on the y axis.</br>

To update the plots with the currently selected parameter combination set on the left bar, please press the calculate button on the left side.</br>

The power analysis can be specific exactly for the specific use case with a lot of different parameters. However, for all parameters, default values are given which can be used. The parameters are divided into different categories.</br>
<h5>General parameters</h5>
<ul>
    <li>Study type: DE or eQTL study.</li>
    <li>Cell type: for the cell-type specific analysis. The expression distribution is selected for this cell type. Only one cell type at once can be analysed. If multiple cell types are of interest (which is often the case), try different cell types, focusing especially on the ones with small cell type frequencies.</li>
    <li>Cell type frequency: of the target cell type.</li>
    <li>Reference study: effect sizes and expression ranks are taken from a reference study, performed on FACS sorted bulk RNA-seq data. Different examples are available for DE as well as eQTL studies.</li>
    <li>Total budget: of the experiment.</li>
    <li>Parameter grid: all possible combinations for two of the three experimental parameters (sample size, cells per person and read depth) are tested, the third parameter is defined uniquely given the other two and the overall budget. Which of the two shall be tested, can be selcted here. Depending on the selection, the four parameters below are adapted.</li>
    <li>"Variable" (min) and "Variable" (max): the tested parameter grid contains always two of the experimental parameters (sample size, cells per person and read dpeth), which ones are shown depends on the parameter selection in "Parameter grid". The input fields define the minimal and maximal value of the tested ranges for the respective parameter.</li>
    <li>Steps: number of values in the parameter ranges for the parameter grid.</li>
</ul>
<h5>Cost and experimental parameters</h5>
The cost function influences how many parameters can be measured for the given overall budget.
<ul>
    <li>Cost 10X kit: Cost for one 10X Genomics kit.</li>
    <li>Cost flow cell: Cost for one flow cell.</li>
    <li>Number of reads per flow cell.</li>
    <li>Cells per lane: Number of cells meassured on one 10X lane, assuming 6 lanes per kit.</li>
</ul>
<h5>Multiple testing correction</h5>
Both the p-value and the multiple testing strategy can be chosen. We recommend using FWER adjustement for eQTL studies and FDR adjustment for DE studies.
<ul>
    <li>P-value: Significance threshold.</li>
    <li>Multiple testing method: Possible is adjustment after the family-wise error rate (FWER), after the false discovery rate (FDR) or no adjustment at all (none).</li>
</ul>
<h5>Mapping and multiplet estimation</h5>
The more cells are loaded on a lane, the more multiplets are produced. These need to be discared before the analysis. Furthermore, the reads are not evenly distributed, but multiplets have a higher fraction of reads per cell than singlets, so the target read depth is reduced with more multiplet.
<ul>
    <li>Mapping efficiency</li>
    <li>Multiplet rate: Rate factor to calculate the number of multiplets dependent on the number of cells loaded per lane. We assume a linear relationship of multiplet fraction = cells per lane * multiplet rate.</li>
    <li>Multiplet factor: multiplets have a higher fraction of reads per cell than singlets, the multiplet factor shows the ratio between the reads.</li>
</ul>
<h5>Expression cutoffs</h5>
A gene is defined as expressed, if it has a certain fraction of UMI counts per gene in a certain fraction of individuals (both parameters can be set). This influences the expression probability.
<ul>
    <li>Minimal number of UMI per gene.</li>
    <li>Fraction of individuals.</li>
</ul>
<h5>Special parameters</h5>
The method of power calculation can be changed to speed calculation or to increase accuracy (especially important for eQTL calculation).
<ul>
    <li>Skip power for lowly expressed genes: this parameter will speed the calculation by setting the power of lowly expressed genes (probability smaller than 0.01) automatically to 0. This will have only a marginal effect on the overall power, but of course change the DE/eQTL power estimates.</li>
    <li>Use simulated power for eQTLs: for genes with small mean values, the method used for eQTL power calculation can get inaccurate. Instead the eQTL power for these small mean values can be estimated via simulation, which however increases the runtime.</li>
</ul>

<h4>Power to detect cell types</h4>
"Power to detect cell types" describes the power to detect a sufficient number of cells from a cell type of interest in each individual. This is important as a cell-type specific DE or eQTL analysis is only possible if enough cells of this cell type are detected. The method calculates the minimal number of cells per individual which are necessary to reach a sufficient power threshold. The necessary parameters, which can be set on the left side, are:

<ul>
    <li>Samples: number of samples of the study.</li>
    <li>Cell type frequency: for the cell type of interest.</li>
    <li>Detection power: the target power that should be reached with the parameter combination.</li>
    <li>Minimal number of cells: how many cells of the target cell type should at least be detected for each individual.</li>
</ul>

<h4>References</h4>

A detailed description of the complete model can be found in our paper</br></br>

"Design and power analysis for multi-sample single cell genomics experiments", Schmid et al., submitted</br></br>

All code including an offline version of this website, build with R shiny, is available as an R package on
<a href="https://github.com/heiniglab/scPower">Github</a>. With the R package, the user can also fit and incorporate own priors for expression probabilities and effect sizes. This is due to runtime reasons not possible over the webserver. The package contains a detailed introduction vignette explaining all necessary steps for the inclusion of own priors.
</br>

